File: common/src/main/java/net/shiroha233/roadweaver/client/map/data/MapDataCollector.java
/* 文件职责：汇总结构、连接与道路数据，生成地图快照。 */
package net.shiroha233.roadweaver.client.map.data;

import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.shiroha233.roadweaver.core.model.StructureConnection;
import net.shiroha233.roadweaver.core.model.StructureInfo;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * 地图快照装配器。
 */
public final class MapDataCollector {
    private MapDataCollector() {}

    public static final int MAX_SNAPSHOT_STRUCTURES = 2048;
    public static final int MAX_SNAPSHOT_CONNECTIONS = 4096;
    public static final int MAX_SNAPSHOT_ROAD_POLYLINES = 4096;
    public static final int MAX_SNAPSHOT_ROAD_POINTS = 262144;
    public static final int INCREMENTAL_RESPONSE_COUNT = 3;

    public static MapSnapshot build(ServerLevel level, int minBlockX, int minBlockZ, int maxBlockX, int maxBlockZ) {
        MapStructureCollector.Result structureResult = MapStructureCollector.collect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ);
        List<StructureConnection> connections = MapConnectionCollector.collect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ);
        List<List<BlockPos>> roads = MapRoadCollector.collect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ);
        return composeSnapshot(structureResult, connections, roads);
    }

    public static MapSnapshot buildStructuresSnapshot(ServerLevel level,
                                                      int minBlockX,
                                                       int minBlockZ,
                                                       int maxBlockX,
                                                       int maxBlockZ) {
        MapStructureCollector.Result structureResult = MapStructureCollector.collect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ);
        List<BlockPos> structures = limitList(structureResult.structures(), MAX_SNAPSHOT_STRUCTURES);
        return new MapSnapshot(
                structures,
                List.of(),
                matchingStructureInfos(structures, structureResult.infos()),
                List.of(),
                structureResult.sources());
    }

    public static MapSnapshot buildRoadsSnapshot(ServerLevel level,
                                                 int minBlockX,
                                                  int minBlockZ,
                                                  int maxBlockX,
                                                  int maxBlockZ) {
        List<List<BlockPos>> roads = MapRoadCollector.collect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ);
        return new MapSnapshot(List.of(), List.of(), List.of(), limitRoadPolylines(roads));
    }

    public static MapSnapshot buildConnectionsSnapshot(ServerLevel level,
                                                       int minBlockX,
                                                        int minBlockZ,
                                                        int maxBlockX,
                                                        int maxBlockZ) {
        List<StructureConnection> connections = MapConnectionCollector.collect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ);
        return new MapSnapshot(List.of(), limitList(connections, MAX_SNAPSHOT_CONNECTIONS), List.of(), List.of());
    }

    private static MapSnapshot composeSnapshot(MapStructureCollector.Result structureResult,
                                               List<StructureConnection> connections,
                                               List<List<BlockPos>> roads) {
        List<BlockPos> structures = limitList(structureResult.structures(), MAX_SNAPSHOT_STRUCTURES);
        return new MapSnapshot(
                structures,
                limitList(connections, MAX_SNAPSHOT_CONNECTIONS),
                matchingStructureInfos(structures, structureResult.infos()),
                limitRoadPolylines(roads),
                structureResult.sources());
    }

    private static List<StructureInfo> matchingStructureInfos(List<BlockPos> structures, List<StructureInfo> infos) {
        if (structures == null || structures.isEmpty() || infos == null || infos.isEmpty()) return List.of();
        HashMap<Long, StructureInfo> byPos = new HashMap<>();
        for (StructureInfo info : infos) {
            if (info != null && info.pos() != null) byPos.put(posKey(info.pos()), info);
        }
        ArrayList<StructureInfo> out = new ArrayList<>(structures.size());
        for (BlockPos structure : structures) {
            StructureInfo info = byPos.get(posKey(structure));
            if (info != null) out.add(info);
        }
        return out;
    }

    private static long posKey(BlockPos pos) {
        return (((long) pos.getX()) << 32) ^ (pos.getZ() & 0xffffffffL);
    }

    private static <T> List<T> limitList(List<T> input, int maxSize) {
        if (input == null || input.isEmpty()) return List.of();
        int limit = Math.max(0, maxSize);
        if (input.size() <= limit) return input;
        return new ArrayList<>(input.subList(0, limit));
    }

    private static List<List<BlockPos>> limitRoadPolylines(List<List<BlockPos>> roads) {
        if (roads == null || roads.isEmpty()) return List.of();
        ArrayList<List<BlockPos>> out = new ArrayList<>();
        int remainingPoints = MAX_SNAPSHOT_ROAD_POINTS;
        for (List<BlockPos> road : roads) {
            if (out.size() >= MAX_SNAPSHOT_ROAD_POLYLINES || remainingPoints <= 1) break;
            if (road == null || road.size() < 2) continue;
            int take = Math.min(road.size(), remainingPoints);
            if (take >= 2) {
                out.add(take == road.size() ? road : new ArrayList<>(road.subList(0, take)));
                remainingPoints -= take;
            }
        }
        return out;
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/client/map/data/MapDataCollector.kt
package net.shiroha233.roadweaver.client.map.data

import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level
import net.shiroha233.roadweaver.config.ConfigService
import net.shiroha233.roadweaver.helpers.Records
import net.shiroha233.roadweaver.persistence.WorldDataProvider
import net.shiroha233.roadweaver.persistence.sharded.RoadShardStorage
import net.shiroha233.roadweaver.planning.RoadPlanningService
import net.shiroha233.roadweaver.search.StructureIndexService
import java.util.ArrayList
import java.util.HashSet

object MapDataCollector {
    // 视图跨度超过该阈值时，不再加载详细道路几何，只依赖连接直线
    private const val MAX_DETAILED_ROAD_SPAN_BLOCKS = 7680

    @JvmStatic
    fun build(level: ServerLevel): MapSnapshot {
        val provider = WorldDataProvider.getInstance()
        val loc = provider.getStructureLocations(level)
        val connections = provider.getStructureConnections(level)

        val structures: MutableList<BlockPos> = if (loc != null) ArrayList(loc.structureLocations) else ArrayList()
        val conns: MutableList<Records.StructureConnection> = if (connections != null) ArrayList(connections) else ArrayList()
        val infos: MutableList<Records.StructureInfo> = if (loc != null) ArrayList(loc.structureInfos) else ArrayList()

        val roads: MutableList<List<BlockPos>> = ArrayList()
        val cfg = ConfigService.get()
        val spawn = level.sharedSpawnPos
        val radiusChunks = kotlin.math.max(1, cfg.initialPlanRadiusChunks())
        val minX = ((spawn.x shr 4) - radiusChunks) * 16
        val maxX = ((spawn.x shr 4) + radiusChunks) * 16
        val minZ = ((spawn.z shr 4) - radiusChunks) * 16
        val maxZ = ((spawn.z shr 4) + radiusChunks) * 16

        val roadDataList = RoadShardStorage.queryRect(level, minX, minZ, maxX, maxZ)
        for (rd in roadDataList) {
            val segs = rd.roadSegmentList
            if (segs.isNullOrEmpty()) continue
            val poly = ArrayList<BlockPos>(segs.size)
            for (sp in segs) poly.add(sp.middlePos)
            if (poly.size >= 2) roads.add(poly)
        }

        if (Level.OVERWORLD == level.dimension()) {
            val verified = StructureIndexService.predictAndVerifyAroundSpawn(level)
            if (verified.isNotEmpty()) {
                val existing = HashSet(structures)
                for (info in verified) {
                    val p = info.pos
                    if (!existing.contains(p)) {
                        structures.add(p)
                        infos.add(info)
                        existing.add(p)
                    }
                }
            }
        }

        return MapSnapshot(structures, conns, infos, roads)
    }

    @JvmStatic
    fun build(level: ServerLevel, minBlockX: Int, minBlockZ: Int, maxBlockX: Int, maxBlockZ: Int): MapSnapshot {
        val provider = WorldDataProvider.getInstance()
        val loc = provider.getStructureLocations(level)
        val connections = provider.getStructureConnections(level)

        val structures = ArrayList<BlockPos>()
        if (loc != null && loc.structureLocations != null) {
            for (p in loc.structureLocations) {
                val x = p.x
                val z = p.z
                if (x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ) structures.add(p)
            }
        }

        val conns = ArrayList<Records.StructureConnection>()
        if (connections != null) {
            for (c in connections) {
                val a = c.from
                val b = c.to
                val ina = a.x >= minBlockX && a.x <= maxBlockX && a.z >= minBlockZ && a.z <= maxBlockZ
                val inb = b.x >= minBlockX && b.x <= maxBlockX && b.z >= minBlockZ && b.z <= maxBlockZ
                if (ina || inb) conns.add(c)
            }
        }

        val infos = ArrayList<Records.StructureInfo>()
        if (loc != null && loc.structureInfos != null) {
            for (info in loc.structureInfos) {
                val p = info.pos
                val x = p.x
                val z = p.z
                if (x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ) infos.add(info)
            }
        }

        if (Level.OVERWORLD == level.dimension()) {
            val verified = StructureIndexService.predictAndVerifyInRect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ)
            if (verified.isNotEmpty()) {
                val existing = HashSet(structures)
                for (info in verified) {
                    val p = info.pos
                    if (!existing.contains(p)) {
                        val x = p.x
                        val z = p.z
                        if (x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ) {
                            structures.add(p)
                            infos.add(info)
                            existing.add(p)
                        }
                    }
                }
            }
        }

        val roads: MutableList<List<BlockPos>> = ArrayList()
        val roadDataList = RoadShardStorage.queryRect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ)
        for (rd in roadDataList) {
            val segs = rd.roadSegmentList
            if (segs.isNullOrEmpty()) continue
            val poly = ArrayList<BlockPos>(segs.size)
            for (sp in segs) {
                val p = sp.middlePos
                val x = p.x
                val z = p.z
                if (x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ) poly.add(p)
            }
            if (poly.size >= 2) roads.add(poly)
        }

        return MapSnapshot(structures, conns, infos, roads)
    }

    @JvmStatic
    fun build(
        level: ServerLevel,
        minBlockX: Int,
        minBlockZ: Int,
        maxBlockX: Int,
        maxBlockZ: Int,
        centerX: Int,
        centerZ: Int,
        radiusBlocks: Int
    ): MapSnapshot {
        val provider = WorldDataProvider.getInstance()
        val loc = provider.getStructureLocations(level)
        val connections = provider.getStructureConnections(level)

        val r = kotlin.math.max(0, radiusBlocks)
        val r2 = r.toLong() * r.toLong()
        val inAOI: (Int, Int) -> Boolean = { x, z ->
            if (r2 <= 0) true
            else {
                val dx = (x - centerX).toLong()
                val dz = (z - centerZ).toLong()
                dx * dx + dz * dz <= r2
            }
        }

        val spanX = kotlin.math.abs(maxBlockX - minBlockX)
        val spanZ = kotlin.math.abs(maxBlockZ - minBlockZ)

        // 先构建“已规划端点”集合（任何状态）
        val plannedEndpoints = HashSet<BlockPos>()
        if (connections != null) {
            for (c in connections) {
                plannedEndpoints.add(c.from)
                plannedEndpoints.add(c.to)
            }
        }

        // 计算“已触发规划覆盖”判断
        val cfgAll = ConfigService.get()
        val initialRadiusBlocks = kotlin.math.max(1, cfgAll.initialPlanRadiusChunks()) * 16
        val strideChunks = RoadPlanningService.getStrideTileSizeChunks()
        val dynRadiusChunks = RoadPlanningService.getDynamicPlanRadiusChunks()
        val dynRadiusBlocks = kotlin.math.max(1, dynRadiusChunks) * 16
        val initialR2 = initialRadiusBlocks.toLong() * initialRadiusBlocks.toLong()
        val spawn = level.sharedSpawnPos

        // 预计算已规划 tiles 对应的近似矩形（块坐标）
        val plannedRects = ArrayList<IntArray>()
        val centersMap = RoadPlanningService.getPlannedTileCenters(level)
        for (key in RoadPlanningService.getPlannedTiles(level)) {
            val kx = (key shr 32).toInt()
            val kz = (key and 0xffffffffL).toInt()

            val cval = centersMap[key]
            val centerChunkX: Int
            val centerChunkZ: Int
            if (cval != null) {
                centerChunkX = (cval shr 32).toInt()
                centerChunkZ = (cval and 0xffffffffL).toInt()
            } else {
                centerChunkX = kx * strideChunks + strideChunks / 2
                centerChunkZ = kz * strideChunks + strideChunks / 2
            }

            val cxBlocks = centerChunkX * 16
            val czBlocks = centerChunkZ * 16
            plannedRects.add(
                intArrayOf(
                    cxBlocks - dynRadiusBlocks,
                    czBlocks - dynRadiusBlocks,
                    cxBlocks + dynRadiusBlocks,
                    czBlocks + dynRadiusBlocks
                )
            )
        }

        val inPlannedCoverage: (Int, Int) -> Boolean = { x, z ->
            var ok = false
            if (!ok && Level.OVERWORLD == level.dimension()) {
                val dxs = (x - spawn.x).toLong()
                val dzs = (z - spawn.z).toLong()
                if (dxs * dxs + dzs * dzs <= initialR2) ok = true
            }
            if (!ok) {
                for (rct in plannedRects) {
                    if (x >= rct[0] && x <= rct[2] && z >= rct[1] && z <= rct[3]) {
                        ok = true
                        break
                    }
                }
            }
            ok
        }

        val structures = ArrayList<BlockPos>()
        if (loc != null && loc.structureLocations != null) {
            for (p in loc.structureLocations) {
                val x = p.x
                val z = p.z
                val inRect = x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ
                if (!inRect) continue

                // 规则：已规划端点始终显示；已触发规划覆盖内的结构显示；其余仅在 AOI 内显示
                if (plannedEndpoints.contains(p) || inPlannedCoverage(x, z) || inAOI(x, z)) {
                    structures.add(p)
                }
            }
        }

        // 确保连接端点一定出现在结构点列表中
        if (connections != null) {
            val existing = HashSet(structures)
            for (c in connections) {
                val eps = arrayOf(c.from, c.to)
                for (ep in eps) {
                    val x = ep.x
                    val z = ep.z
                    val inRect = x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ
                    if (inRect && !existing.contains(ep)) {
                        structures.add(ep)
                        existing.add(ep)
                    }
                }
            }
        }

        val conns = ArrayList<Records.StructureConnection>()
        if (connections != null) {
            for (c in connections) {
                val a = c.from
                val b = c.to
                val ina = a.x >= minBlockX && a.x <= maxBlockX && a.z >= minBlockZ && a.z <= maxBlockZ
                val inb = b.x >= minBlockX && b.x <= maxBlockX && b.z >= minBlockZ && b.z <= maxBlockZ
                // 规则：连接按矩形过滤，不受 AOI 限制
                if (ina || inb) conns.add(c)
            }
        }

        val infos = ArrayList<Records.StructureInfo>()
        if (loc != null && loc.structureInfos != null) {
            for (info in loc.structureInfos) {
                val p = info.pos
                val x = p.x
                val z = p.z
                val inRect = x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ
                if (!inRect) continue

                if (plannedEndpoints.contains(p) || inPlannedCoverage(x, z) || inAOI(x, z)) {
                    infos.add(info)
                }
            }
        }

        val roads: MutableList<List<BlockPos>> = ArrayList()

        // 视图跨度太大时跳过道路几何的加载，只用连接直线表示已连接道路
        val loadDetailedRoads = spanX <= MAX_DETAILED_ROAD_SPAN_BLOCKS && spanZ <= MAX_DETAILED_ROAD_SPAN_BLOCKS
        if (loadDetailedRoads) {
            val roadDataList = RoadShardStorage.queryRect(level, minBlockX, minBlockZ, maxBlockX, maxBlockZ)
            for (rd in roadDataList) {
                val segs = rd.roadSegmentList
                if (segs.isNullOrEmpty()) continue

                val poly = ArrayList<BlockPos>(segs.size)
                for (sp in segs) {
                    val p = sp.middlePos
                    val x = p.x
                    val z = p.z
                    if (x >= minBlockX && x <= maxBlockX && z >= minBlockZ && z <= maxBlockZ) {
                        poly.add(p)
                    }
                }

                if (poly.size >= 2) roads.add(poly)
            }
        }

        return MapSnapshot(structures, conns, infos, roads)
    }
}
